Mathlib/CategoryTheory/Yoneda.lean

```diff
@@ -543,4 +544,201 @@ lemma isIso_of_yoneda_map_bijective {X Y : C} (f : X ⟶ Y)
+@[ext]
+lemma coyonedaPairingExt {X : C × (C ⥤ Type v₁)} {x y : (coyonedaPairing C).obj X}
+    (w : ∀ Y, x.app Y = y.app Y) : x = y :=
+  NatTrans.ext _ _ (funext w)
+
+@[simp]
+theorem coyonedaPairing_map (P Q : C × (C ⥤ Type v₁)) (α : P ⟶ Q) (β : (coyonedaPairing C).obj P) :
+    (coyonedaPairing C).map α β = coyoneda.map α.1.op ≫ β ≫ α.2 :=
+  rfl
+
+universe w in
+variable {C} in
+/-- A bijection `(coyoneda.obj X ⋙ uliftFunctor ⟶ F) ≃ F.obj (unop X)` which is a variant
+of `coyonedaEquiv` with heterogeneous universes. -/
+def coyonedaCompUliftFunctorEquiv (F : C ⥤ Type max v₁ w) (X : Cᵒᵖ) :
+    (coyoneda.obj X ⋙ uliftFunctor.{w} ⟶ F) ≃ F.obj X.unop where
+  toFun φ := φ.app X.unop (ULift.up (𝟙 _))
+  invFun f :=
+    { app := fun Y x => F.map (ULift.down x) f }
+  left_inv φ := by
+    ext Y f
+    dsimp
+    rw [← FunctorToTypes.naturality]
+    dsimp
+    rw [Category.id_comp]
+    rfl
+  right_inv f := by aesop_cat
+
+/-- The Coyoneda lemma asserts that the Coyoneda pairing
+`(X : C, F : C ⥤ Type) ↦ (coyoneda.obj X ⟶ F)`
+is naturally isomorphic to the evaluation `(X, F) ↦ F.obj X`.
+
+See <https://stacks.math.columbia.edu/tag/001P>.
+-/
+def coyonedaLemma : coyonedaPairing C ≅ coyonedaEvaluation C :=
+  NatIso.ofComponents
+    (fun X ↦ Equiv.toIso (coyonedaEquiv.trans Equiv.ulift.{u₁, v₁}.symm))
+    (by intro (X, F) (Y, G) f
+        ext (a : coyoneda.obj (op X) ⟶ F)
+        apply ULift.ext.{u₁, v₁}
+        simp only [Functor.prod_obj, Functor.id_obj, types_comp_apply, coyonedaEvaluation_map_down]
+        erw [Equiv.ulift_symm_down, Equiv.ulift_symm_down]
+        simp [coyonedaEquiv, ← FunctorToTypes.naturality])
+
+variable {C}
+
+/-- The isomorphism between `coyoneda.obj X ⟶ F` and `F.obj (unop X)`
+(we need to insert a `ULift` to get the universes right!)
+given by the Coyoneda lemma.
+-/
+@[simps!]
+def coyonedaSections (X : Cᵒᵖ) (F : C ⥤ Type v₁) :
+    (coyoneda.obj X ⟶ F) ≅ ULift.{u₁} (F.obj X.unop) :=
+  (coyonedaLemma C).app (X.unop, F)
+
+/-- When `C` is a small category, we can restate the isomorphism from `coyonedaSections`
+without having to change universes.
+-/
+def coyonedaSectionsSmall {C : Type u₁} [SmallCategory C] (X : Cᵒᵖ) (F : C ⥤ Type u₁) :
+    (coyoneda.obj X ⟶ F) ≅ F.obj X.unop :=
+  coyonedaSections X F ≪≫ uliftTrivial _
+
+@[simp]
+theorem coyonedaSectionsSmall_hom {C : Type u₁} [SmallCategory C] (X : Cᵒᵖ) (F : C ⥤ Type u₁)
+    (f : coyoneda.obj X ⟶ F) : (coyonedaSectionsSmall X F).hom f = f.app _ (𝟙 _) :=
+  rfl
+
+@[simp]
+theorem coyonedaSectionsSmall_inv_app_apply {C : Type u₁} [SmallCategory C] (X : Cᵒᵖ)
+    (F : C ⥤ Type u₁) (t : F.obj (unop X)) (Y : C) (f : X.unop ⟶ Y) :
+    ((coyonedaSectionsSmall X F).inv t).app Y f = F.map f t :=
+  rfl
+
+attribute [local ext] Functor.ext
+
+/- Porting note: this used to be two calls to `tidy` -/
+/-- The curried version of coyoneda lemma when `C` is small. -/
+def curriedCoyonedaLemma {C : Type u₁} [SmallCategory C] :
```

Comment:
The definition and the next one does not need `ulift`, so that I believe it should be defined directly as:
```lean
/-- The curried version of coyoneda lemma when `C` is small. -/
def curriedCoyonedaLemma {C : Type u₁} [SmallCategory C] :
    (coyoneda.rightOp ⋙ coyoneda : C ⥤ (C ⥤ Type u₁) ⥤ Type u₁) ≅ evaluation C (Type u₁) :=
  NatIso.ofComponents (fun X =>
    NatIso.ofComponents (fun F => Equiv.toIso coyonedaEquiv)) (by sorry)

/-- The curried version of coyoneda lemma when `C` is small. -/
def curriedCoyonedaLemma' {C : Type u₁} [SmallCategory C] :
    yoneda ⋙ (whiskeringLeft C (C ⥤ Type u₁)ᵒᵖ (Type u₁)).obj coyoneda.rightOp
      ≅ 𝟭 (C ⥤ Type u₁) :=
  NatIso.ofComponents (fun F => NatIso.ofComponents (fun X => Equiv.toIso coyonedaEquiv) sorry)
```
(By extending `Functor.flip` to make it a functor which sends a functor to its flip, it should be possible to define the curried iso from the first, but I am not sure there it a strong need from this API for bifunctors.)